File: src/velvet/util/structs/AutoSet.java
package velvet.util.structs;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Optional;
import java.util.Random;
import java.util.function.Function;
import java.util.stream.Stream;

/*
Functions as a set for a list of items
The auto-set automatically extracts a key from each item, and uses this key to store the value
Essentially functions as a normal set, with an override on the key extractor
 */
public class AutoSet<K, V> {

    private static final Random random = new Random();

    private final Function<V, K> keyExtractor;
    private final ArrayList<K> keys;
    private final HashMap<K, V> map;

    public AutoSet(Function<V, K> keyExtractor) {
        this.keyExtractor = keyExtractor;
        keys = new ArrayList<>();
        map = new HashMap<>();
    }

    public void add(V value){
        K key = keyExtractor.apply(value);
        if(map.containsKey(key)){ return; }

        keys.add(key);
        map.put(key, value);
    }

    public Optional<V> get(K key){
        return Optional.ofNullable(map.get(key));
    }

    public V random(){ return map.get(keys.get(random.nextInt(keys.size()))); }

    public boolean contains(V value){
        return map.containsKey(keyExtractor.apply(value));
    }

    public int size(){ return map.size(); }

    public void clear(){
        keys.clear();
        map.clear();
    }

    public void remove(V value){
        K key = keyExtractor.apply(value);
        if(!map.containsKey(key)){ return; }

        keys.remove(key);
        map.remove(key);
    }

    public Stream<V> getValues(){
        return keys.stream().map(map::get);
    }
}

File: src/velvet/util/structs/AutoSet.kt
package velvet.util.structs

import java.util.*
import java.util.function.Function
import java.util.stream.Stream
import kotlin.random.Random

class AutoSet<K, V>(private val keyExtractor: (V)->K) {

    private val map: MutableMap<K, V> = mutableMapOf()

    val size: Int
        get() = map.size
    val values: Sequence<V>
        get() = map.asSequence().map { it.value }

    fun add(value: V) {
        val key = keyExtractor(value)
        if (key in map) {
            return
        }
        map[key] = value
    }

    operator fun get(key: K): V? = map[key]

    operator fun contains(value: V) = keyExtractor(value) in map

    fun clear() {
        map.clear()
    }

    fun remove(value: V) {
        map.remove(keyExtractor(value))
    }
}